gitlab-runner/src/client.rs

```diff
@@ -275,20 +282,72 @@ pub(crate) struct Client {
     client: reqwest::Client,
     url: Url,
     token: String,
+    system_id: String,
 }
 
 impl Client {
     pub fn new(url: Url, token: String) -> Self {
+        let system_id = Self::generate_system_id();
         Self {
             client: reqwest::Client::new(),
             url,
             token,
+            system_id,
+        }
+    }
+
+    fn generate_system_id_from_machine_id() -> Option<String> {
+        // Ideally this would be async for consistency, but that's for the next API bump really. In
+        // practise the client will be created at the start of the runner and the amount read is
+        // really tiny so blocking is not a real issue.
+        let mut f = match File::open("/etc/machine-id") {
+            Ok(f) => f,
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return None,
+            Err(e) => {
+                warn!("Failed to open machine-id: {e}");
+                return None;
+            }
+        };
+
+        let mut id = [0u8; 32];
+        match f.read(&mut id) {
+            Ok(r) if r != 32 => {
+                warn!("Short read from machine-id (only {} bytes)", r);
+                return None;
+            }
+            Err(e) => {
+                warn!("Failed to read from machine-id: {e}");
+                return None;
+            }
+            _ => (),
+        };
+
+        // Infallable as a hmac can take a key of any size
+        let mut mac = Hmac::<sha2::Sha256>::new_from_slice(&id).unwrap();
+        mac.update(b"gitlab-runner");
+
+        let mut system_id = String::from("s_");
+        for b in &mac.finalize().into_bytes()[0..6] {
+            // Infallable: writing to a string
+            write!(&mut system_id, "{:02x}", b).unwrap();
+        }
+        Some(system_id)
+    }
+
+    fn generate_system_id() -> String {
+        if let Some(system_id) = Self::generate_system_id_from_machine_id() {
+            system_id
+        } else {
+            let mut system_id = String::from("r_");
+            Alphanumeric.append_string(&mut rand::thread_rng(), &mut system_id, 12);
```

Comment:
Perhaps make `12` a shared constant, since it's the origin of the 6-byte(=12 hex char) limit above too?